### PR TITLE
Implement tchannel/as/json

### PR DESCRIPTION
**Pull request by Raynos**
This is a first pass implementation of `TChannelJSON` as per the Overview.md
document.

It's a little heavy but handles the majority of the concerns.

r: @jcorbin @kriskowal

### CODE REVIEW

**Comment by kriskowal** [comment]
:ok:

File: node/as/json.js [hunk 1]
```diff
@@ -0,0 +1,342 @@
+'use strict';
```
Comment (jcorbin) [inline node/as/json.js]: missing license

File: node/as/json.js [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+var Buffer = require('buffer').Buffer;
+var assert = require('assert');
+var NullLogtron = require('null-logtron');
+var Result = require('bufrw/result');
+var cyclicStringify = require('json-stringify-safe');
+var WrappedError = require('error/wrapped');
+var TypedError = require('error/typed');
+var isError = require('is-error');
+
+var HeadParserError = WrappedError({
```
Comment (jcorbin) [inline node/as/json.js]: move all these error decls into errors module

File: node/as/json.js [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        'req.service must be a string');
+    assert(body !== undefined, 'must send a body');
+
+    var stringifyResult = self._stringify({
+        head: head,
+        body: body,
+        endpoint: endpoint,
+        direction: 'out.request'
+    });
+    if (stringifyResult.error) {
```
Comment (jcorbin) [inline node/as/json.js]: .err?

File: node/as/json.js [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+            direction: opts.direction,
+            head: cyclicStringify(opts.head)
+        });
+
+        self.logger.error('Got unexpected unserializable JSON for arg2', {
+            endpoint: opts.endpoint,
+            direction: opts.direction,
+            headErr: headStringifyErr
+        });
+        return new Result(headStringifyErr);
```
Comment (jcorbin) [inline node/as/json.js]: return Result.error(err)

File: node/as/json.js [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    var bodyR = safeJSONStringify(opts.body);
+    if (bodyR.error) {
```
Comment (jcorbin) [inline node/as/json.js]: .err

File: node/as/json.js [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+            );
+        }
+    }
+};
+
+TChannelJSON.prototype._stringify = function stringify(opts) {
+    var self = this;
+
+    var headR = safeJSONStringify(opts.head);
+    if (headR.error) {
```
Comment (jcorbin) [inline node/as/json.js]: .err

File: node/as/json.js [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+        var headStringifyErr = HeadStringifyError(headR.error, {
```
Comment (jcorbin) [inline node/as/json.js]: .err

File: node/as/json.js [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+            endpoint: opts.endpoint,
+            direction: opts.direction,
+            headErr: headStringifyErr
+        });
+        return new Result(headStringifyErr);
+    }
+
+    var bodyR = safeJSONStringify(opts.body);
+    if (bodyR.error) {
+        var bodyStringifyErr = BodyStringifyError(bodyR.error, {
```
Comment (jcorbin) [inline node/as/json.js]: .err

File: node/as/json.js [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+        head: headR.value,
+        body: bodyR.value
+    });
+};
+
+TChannelJSON.prototype._parse = function parse(opts) {
+    var self = this;
+
+    var headR = safeJSONParse(opts.head);
+    if (headR.error) {
```
Comment (jcorbin) [inline node/as/json.js]: .err... and such

File: node/as/json.js [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+
+    // jscs:disable
+    try {
+        json = JSON.parse(str);
+    } catch (e) {
+        return new Result(e);
+    }
+    // jscs:enable
+
+    return new Result(null, json);
```
Comment (jcorbin) [inline node/as/json.js]: = null default?

File: node/as/json.js [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+function TChannelJSONResponse(ok, head, body) {
+    var self = this;
+
+    self.ok = ok;
+    self.head = head;
+    self.body = body;
+}
+
+function safeJSONStringify(obj) {
+    var str;
```
Comment (jcorbin) [inline node/as/json.js]: = '' default

**Pull request merged**